sys/include/ztimer.h

```diff
@@ -0,0 +1,328 @@
+ * @param[in]   val         alarm target
+ */
+void ztimer_set(ztimer_dev_t *ztimer, ztimer_t *entry, uint32_t val);
+
+/**
+ * @brief   Remove an alarm from a clock
+ *
+ * This will place @p entry in the timer targets queue for @p ztimer.
+ *
+ * This function does nothing if @p entry is not found in the alarm queue of @p ztimer
+ *
+ * @param[in]   ztimer      ztimer clock to operate on
+ * @param[in]   entry       alarm entry to enqueue
+ * @param[in]   val         alarm target
+ */
+void ztimer_remove(ztimer_dev_t *ztimer, ztimer_t *entry);
+
+/**
+ * @brief   Post a message after a delay
+ *
+ * This function sets an alarm that will send a message @p offset ticks
+ * from now.
+ *
+ * @note The memory pointed at by @p alarm and @p msg will not be copied, i.e.
+ *       `*alarm` and `*msg` needs to remain valid until the alarm has occurred.
+ *
+ * @param[in]   ztimer          ztimer clock to operate on
+ * @param[in]   alarm           pointer to alarm struct
+ * @param[in]   offset          ticks from now
+ * @param[in]   msg             pointer to msg that will be sent
+ * @param[in]   target_pid      pid the message will be sent to
+ */
+void ztimer_set_msg(ztimer_dev_t *dev, ztimer_t *timer, uint32_t offset,
+                    msg_t *msg, kernel_pid_t target_pid);
+
+/**
+ * @brief receive a message blocking, with timeout
+ *
+ * @param[in]   ztimer          ztimer clock to operate on
+ * @param[out]  msg             pointer to buffer which will be filled if a message is received
+ * @param[in]   timeout         relative timeout, in @p ztimer time units
+ *
+ * @return  >=0 if a message was received
+ * @return  <0 on timeout
+ */
+int ztimer_msg_receive_timeout(ztimer_dev_t *dev, msg_t *msg, uint32_t timeout);
+
+/**
+ * @brief   Get the current time from a clock
+ *
+ * @param[in]   ztimer          ztimer clock to operate on
+ *
+ * @return  Current count on the clock @p ztimer
+ */
+static inline uint32_t ztimer_now(ztimer_dev_t *ztimer)
+{
+    return ztimer->ops->now(ztimer);
+}
+
+/**
+ * @brief Suspend the calling thread until the time (@p last_wakeup + @p period)
+ *
+ * This function can be used to create periodic wakeups.
+ *
+ * When the function returns, @p last_wakeup is set to
+ * (@p last_wakeup + @p period).
+ *
+ * @c last_wakeup should be set to ztimer_now(@p ztimer) before first call of the
+ * function.
+ *
+ * If the time (@p last_wakeup + @p period) has already passed, the function
+ * sets @p last_wakeup to @p last_wakeup + @p period and returns immediately.
+ *
+ * @param[in]   ztimer          ztimer clock to operate on
+ * @param[in]   last_wakeup     base time stamp for the wakeup
+ * @param[in]   period          time in ticks that will be added to @p last_wakeup
+ */
+void ztimer_periodic_wakeup(ztimer_dev_t *ztimer, uint32_t *last_wakeup,
+                            uint32_t period);
+
+/**
+ * @brief   Put the calling thread to sleep for the specified number of ticks
```

Comment:
Probably also valid for the other functions: This probably doesn't sleep exactly the number of ticks (or seconds) on every platform. Does this sleep at least `ticks` or at most `ticks`. Furthermore, is there a way to get the other behaviour, for example, sleep at least $duration to allow for a peripheral to initialize or sleep at most $duration to not miss a deadline.

Reply:
I bet that this is impossible, as e.g. a thread with higher priority might get the CPU at the very moment the caller wants to wake up. As far as I know, RIOT never had the goal to provide **hard** real time features, but **soft** real time features instead.

Reply:
> Does this sleep at least `ticks` or at most `ticks`.

Well, the implementation sleeps "at least x ticks". I guess this information has to be added everywhere.

> As far as I know, RIOT never had the goal to provide **hard** real time features, but **soft** real time features instead.

I think that is what the website states. Actually, it is more complicated. Parts of RIOT are fully hard real-time capable, some parts are not, some parts are only if certain constraints are honored. With care, it is perfectly possible to do hard real time with RIOT. **but**, you really have to know the internals, as documentation on the matter is severley lacking.
Still, I like that doing hard real-time with RIOT is possible.